components/canvas/canvas_msg.rs

```diff
@@ -0,0 +1,46 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use canvas_paint_task::FillOrStrokeStyle;
+use geom::matrix2d::Matrix2D;
+use geom::point::Point2D;
+use geom::rect::Rect;
+use geom::size::Size2D;
+use std::sync::mpsc::{Sender};
+
+#[derive(Clone)]
+pub enum CanvasMsg {
+    // 2dContext
+    FillRect(Rect<f32>),
+    ClearRect(Rect<f32>),
+    StrokeRect(Rect<f32>),
+    BeginPath,
+    ClosePath,
+    Fill,
+    Stroke,
+    DrawImage(Vec<u8>, Size2D<f64>, Rect<f64>, Rect<f64>, bool),
+    DrawImageSelf(Size2D<f64>, Rect<f64>, Rect<f64>, bool),
+    MoveTo(Point2D<f32>),
+    LineTo(Point2D<f32>),
+    QuadraticCurveTo(Point2D<f32>, Point2D<f32>),
+    BezierCurveTo(Point2D<f32>, Point2D<f32>, Point2D<f32>),
+    Arc(Point2D<f32>, f32, f32, f32, bool),
+    ArcTo(Point2D<f32>, Point2D<f32>, f32),
+    SetFillStyle(FillOrStrokeStyle),
+    SetStrokeStyle(FillOrStrokeStyle),
+    SetLineWidth(f32),
+    SetMiterLimit(f32),
+    SetGlobalAlpha(f32),
+    SetTransform(Matrix2D<f32>),
+    Recreate(Size2D<i32>),
+    GetImageData(Rect<f64>, Size2D<f64>, Sender<Vec<u8>>),
+    PutImageData(Vec<u8>, Rect<f64>, Option<Rect<f64>>),
+    Close,
+    // WebGL
```

Comment:
Oh, I see what you're doing. Can you split this up into three enums: `Canvas2dMsg`, `CanvasWebGlMsg`, and `CommonCanvasMsg`? Then you can just have `CanvasMsg { Canvas2d(Canvas2dMsg), WebGl(CanvasWebGlMsg), Common(CommonCanvasMsg) }`.